Introduce customelementregistry content attribute and avoid changing null registry when inserting a node

rniwa opened the pull request:
#### 30db0392404836cdf3341bbe233d3e55e5886017
<pre>
Introduce customelementregistry content attribute and avoid changing null registry when inserting a node
<a href="https://bugs.webkit.org/show_bug.cgi?id=302583">https://bugs.webkit.org/show_bug.cgi?id=302583</a>

Reviewed by Chris Dumez.

This PR introduces customelementregistry content attribute on element which sets the custom element registry of an element to null,
It also adds the support for specifying `null` in `customElementRegistry` to use null custom element registry in `document.createElement`,
`document.createElementNS`, and `element.attachShadow`.

In addition, this PR changes the semantics of inserting a node so that inserting an element with null custom element registry will not
update its custom element registry to that of the new parent.

These behavior changes were discussed during TPAC as explained in: <a href="https://github.com/whatwg/dom/issues/1413#issuecomment-3530198733">https://github.com/whatwg/dom/issues/1413#issuecomment-3530198733</a>

Test: imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute.html
      imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-in-xhtml.xhtml
      imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-append-does-not-upgrade.html

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/Element-customElementRegistry-exceptions.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/Element-customElementRegistry.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/ShadowRoot-init-customElementRegistry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/ShadowRoot-init-customElementRegistry.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-in-xhtml-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-in-xhtml.xhtml: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-append-does-not-upgrade-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-append-does-not-upgrade.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-define-upgrade-criteria-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-define-upgrade-criteria.html:
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::constructElementWithFallback): Add the element to scoped custom element registry map if needed.
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::enqueueUpgradeInShadowIncludingTreeOrder): Fixed the bug that this code wasn&apos;t accounting for null custom element registry.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createElementForBindings): Respect `null` being specified in Init dictionary.
(WebCore::Document::createElementNS): Ditto.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged): Mark the element as it uses null custom element registry if parser sets customelementregistry content attribute.
(WebCore::Element::insertedIntoAncestor): Rewrote the logic to update the custom element registry map.
* Source/WebCore/dom/ElementCreationOptions.h:
* Source/WebCore/dom/ElementCreationOptions.idl:
* Source/WebCore/dom/Node.cpp:
(WebCore::adoptCustomElementRegistryIfNotExplicitlySet): Added.
(WebCore::Node::moveShadowTreeToNewDocumentFastCase): Set the custom element registry if a given adopted node uses null registry.
(WebCore::Node::moveShadowTreeToNewDocumentSlowCase): Ditto.
(WebCore::Node::moveNodeToNewDocumentFastCase): Ditto.
(WebCore::Node::moveNodeToNewDocumentSlowCase): Ditto.
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::createHTMLElementOrFindCustomElementInterface): Handle customelementregistry content attribute.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::handleNamespaceAttributes):
(WebCore::handleElementAttributes):
(WebCore::XMLDocumentParser::startElementNs): Handle customelementregistry content attribute.

Canonical link: <a href="https://commits.webkit.org/303300@main">https://commits.webkit.org/303300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee073bd0a6510a73787fe107c3726338b4863713

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4375 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139395 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83771 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b5967a5f-3fc5-4865-9660-66e1aefedab2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133753 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4314 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4137 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100807 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/83e82a88-8639-475c-8511-959479c440ad) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134829 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3083 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118118 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81597 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/878691fe-8539-446f-a454-1f75c1018509) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2971 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/827 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82615 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111717 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36242 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142039 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4044 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36821 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109181 "Found 1 new test failure: fast/scrolling/rtl-scrollbars-sticky-document.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4125 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3543 "Found 2 new test failures: fast/images/individual-animation-toggle.html http/tests/webgpu/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109347 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3074 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114398 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57266 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20519 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4098 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32798 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3930 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67545 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4190 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4058 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3375,9 +3379,14 @@ ExceptionOr<ShadowRoot&> Element::attachShadow(const ShadowRootInit& init, std::
-    RefPtr registry = init.customElementRegistry;
-    if (registry && !registry->isScoped() && registry != document().customElementRegistry())
-        return Exception { ExceptionCode::NotSupportedError };
+    RefPtr<CustomElementRegistry> registry = nullptr;
```
No need for `= nullptr`

rniwa replied:
Fixed.

rniwa committed:
Introduce customelementregistry content attribute and avoid changing null registry when inserting a node
https://bugs.webkit.org/show_bug.cgi?id=302583

Reviewed by Chris Dumez.

This PR introduces customelementregistry content attribute on element which sets the custom element registry of an element to null,
It also adds the support for specifying `null` in `customElementRegistry` to use null custom element registry in `document.createElement`,
`document.createElementNS`, and `element.attachShadow`.

In addition, this PR changes the semantics of inserting a node so that inserting an element with null custom element registry will not
update its custom element registry to that of the new parent.

These behavior changes were discussed during TPAC as explained in: https://github.com/whatwg/dom/issues/1413#issuecomment-3530198733

Test: imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute.html
      imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-in-xhtml.xhtml
      imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-append-does-not-upgrade.html

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/Element-customElementRegistry-exceptions.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/Element-customElementRegistry.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/ShadowRoot-init-customElementRegistry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/ShadowRoot-init-customElementRegistry.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-in-xhtml-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute-in-xhtml.xhtml: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-custom-element-registry-customelementregistry-attribute.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-append-does-not-upgrade-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-append-does-not-upgrade.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-define-upgrade-criteria-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/registries/scoped-registry-define-upgrade-criteria.html:
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::constructElementWithFallback): Add the element to scoped custom element registry map if needed.
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::enqueueUpgradeInShadowIncludingTreeOrder): Fixed the bug that this code wasn't accounting for null custom element registry.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createElementForBindings): Respect `null` being specified in Init dictionary.
(WebCore::Document::createElementNS): Ditto.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged): Mark the element as it uses null custom element registry if parser sets customelementregistry content attribute.
(WebCore::Element::insertedIntoAncestor): Rewrote the logic to update the custom element registry map.
* Source/WebCore/dom/ElementCreationOptions.h:
* Source/WebCore/dom/ElementCreationOptions.idl:
* Source/WebCore/dom/Node.cpp:
(WebCore::adoptCustomElementRegistryIfNotExplicitlySet): Added.
(WebCore::Node::moveShadowTreeToNewDocumentFastCase): Set the custom element registry if a given adopted node uses null registry.
(WebCore::Node::moveShadowTreeToNewDocumentSlowCase): Ditto.
(WebCore::Node::moveNodeToNewDocumentFastCase): Ditto.
(WebCore::Node::moveNodeToNewDocumentSlowCase): Ditto.
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::createHTMLElementOrFindCustomElementInterface): Handle customelementregistry content attribute.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::handleNamespaceAttributes):
(WebCore::handleElementAttributes):
(WebCore::XMLDocumentParser::startElementNs): Handle customelementregistry content attribute.

Canonical link: https://commits.webkit.org/303300@main
